### PR TITLE
feat(node-taint): Adding the node taint experiment

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham.chaudhary@mayadata.io>

- fixes: litmuschaos/litmus#1667

### CODE REVIEW

**Comment by ispeakc0de** [comment]
**Logs**

```
time="2020-07-07T14:58:42Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2020-07-07T14:58:42Z" level=info msg="[PreReq]: Updating the chaos result of node-taint experiment (SOT)"
time="2020-07-07T14:58:42Z" level=info msg="The application informations are as follows" Namespace=shubham Label="run=nginx" Ramp Time=0
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of container are as follows" Pod=nginx-65f65459-d6zdj Status=Running container=nginx
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Verify that the Auxiliary Applications are running (pre-chaos)"
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="Successfully added taint on node shubhamworker1-virtual-machine"
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Verify the status of AUT after reschedule"
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Verify that the Auxiliary Applications are running"
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:58:42Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:58:42Z" level=info msg="[Wait]: Waiting for the 60s chaos duration"
time="2020-07-07T14:59:42Z" level=info msg="Successfully removed taint from the shubhamworker1-virtual-machine node"
time="2020-07-07T14:59:42Z" level=info msg="[Confirmation]: The application node has been tainted successfully"
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:59:42Z" level=info msg="The running status of Pods are as follows" Status=Running Pod=nginx-65f65459-d6zdj
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:59:42Z" level=info msg="The running status of container are as follows" Pod=nginx-65f65459-d6zdj Status=Running container=nginx
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Verify that the Auxiliary Applications are running (post-chaos)"
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Checking whether application pods are in running state"
time="2020-07-07T14:59:42Z" level=info msg="The running status of Pods are as follows" Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:59:42Z" level=info msg="[Status]: Checking whether application containers are in running state"
time="2020-07-07T14:59:42Z" level=info msg="The running status of container are as follows" container=nginx Pod=nginx-65f65459-d6zdj Status=Running
time="2020-07-07T14:59:42Z" level=info msg="[The End]: Updating the chaos result of node-taint experiment (EOT)"

```

File: chaoslib/litmus/node_taint/node-taint.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+//waitForChaosDuration waits for the given chaos duration (in seconds)
+func waitForChaosDuration(experimentsDetails *experimentTypes.ExperimentDetails) {
+	time.Sleep(time.Duration(experimentsDetails.ChaosDuration) * time.Second)
+}
+
+// TaintNode taint the application node
+func TaintNode(experimentsDetails *experimentTypes.ExperimentDetails, clients clients.ClientSets) error {
+
+	// get the taint labels & effect
+	TaintKey, TaintValue, TaintEffect := GetTaintDetails(experimentsDetails)
```
Comment (uditgaurav) [inline chaoslib/litmus/node_taint/node-taint.go:91]: Do you also want to give an `operator` along with the key, value and effect?

File: pkg/generic/node-taint/environment/environment.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	experimentDetails.AppNS = Getenv("APP_NAMESPACE", "")
+	experimentDetails.AppLabel = Getenv("APP_LABEL", "")
+	experimentDetails.AppKind = Getenv("APP_KIND", "")
+	experimentDetails.ChaosUID = clientTypes.UID(Getenv("CHAOS_UID", ""))
+	experimentDetails.InstanceID = Getenv("INSTANCE_ID", "")
+	experimentDetails.ChaosPodName = Getenv("POD_NAME", "")
+	experimentDetails.AuxiliaryAppInfo = Getenv("AUXILIARY_APPINFO", "")
+	experimentDetails.AppNode = Getenv("APP_NODE", "")
+	experimentDetails.TaintLabel = Getenv("TAINT_LABEL", "")
+}
```
Comment (uditgaurav) [inline pkg/generic/node-taint/environment/environment.go:30]: Are we not taking effect as env? should we take that?

File: chaoslib/litmus/node_taint/node-taint.go (hunk 1 above)
Comment (ispeakc0de) [reply chaoslib/litmus/node_taint/node-taint.go:91]: operator is only needed in toleration not in taint. Isn't it?

File: pkg/generic/node-taint/environment/environment.go (hunk 2 above)
Comment (ispeakc0de) [reply pkg/generic/node-taint/environment/environment.go:30]: Not effect & key will be part of `TaintLabel` attribute. Can be provided as `key=value:effect`. Name is misleading, We can have a new name i/e Taints

File: chaoslib/litmus/node_taint/node-taint.go (hunk 1 above)
Comment (uditgaurav) [reply chaoslib/litmus/node_taint/node-taint.go:91]: Oh! Correct this is taint!

File: pkg/generic/node-taint/environment/environment.go (hunk 2 above)
Comment (uditgaurav) [reply pkg/generic/node-taint/environment/environment.go]: Let's also give an example in comment above  in job spec like
key=value:NoSchedule

File: chaoslib/litmus/node_taint/node-taint.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+// GetTaintDetails return the key, value and effect for the taint
+func GetTaintDetails(experimentsDetails *experimentTypes.ExperimentDetails) (string, string, string) {
+	TaintValue := "node-taint"
+	TaintEffect := string(apiv1.TaintEffectNoExecute)
+
+	Taints := strings.Split(experimentsDetails.Taints, ":")
+	TaintLabel := strings.Split(Taints[0], "=")
+	TaintKey := TaintLabel[0]
+
+	if len(TaintLabel) >= 2 {
```
Comment (ksatchit) [inline chaoslib/litmus/node_taint/node-taint.go:175]: comments here please?

**Pull request merged**